GHC 8.4 package overrides

deepfire opened the pull request:
###### Motivation for this change
Supply a set of fixes to make more packages buildable with GHC 8.4-alpha2.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

deepfire commented:
@peti, this PR is not _yet_ intended for merging -- rather to gather your feedback.

Compared to the previous situation this set of overrides has:

1. Trim of every non-essential override.
2. Proof that every remaining property override is necessary -- citing the error messages during build, if the said property override is not provided.

This has been derived 95% automatically -- the remainder (just two overrides) was explained manually.

As I remember the previous conversation (in https://github.com/NixOS/nixpkgs/pull/34023), what remains to be done is:
> (b) known to upstream,

This part only makes sense to satisfy if the fix hasn't been mainlined yet.

Thankfully, the automation I have right now is helpful enough to ease that filtering -- and supply classification otherwise -- if you look at the overrides, they are grouped in three categories: 
1. on Hackage
2. upstreamed
3. not upstreamed

So.. I guess I'd appreciate any feedback you might have : -)

deepfire commented:
The point about upstream tracking is now satisfied: the PR URLs are supplied for those of packages that are not upstreamed.

peti commented on pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix:
```diff
@@ -41,4 +41,1041 @@ self: super: {
+  ## On Hackage:
+
+  blaze-markup = overrideCabal super.blaze-markup (drv: {
+    ## On Hackage, awaiting for import
```
It might be easier to use `self.blaze-markup_0_8_2_0`? The same applies to virtually all those overrides that update a package to the latest version. We *have* the latest version in Nixpkgs, and using those names has the advantage that evaluation will break every time these receive an update, which gives us a nudge to re-evaluate the situation and follow the update (or drop the overrive).

deepfire replied:
@peti, I'm trying to replace the src-based override with these "shadow" attributes:
```
  hspec-core = overrideCabal super.hspec-core_2_4_7 (drv: {
    editedCabalFile = null;
  });
```
..and I'm getting:
```
FATAL:  failed to instantiate text-lens:  nix-instantiate packages.nix -A text-lens --argstr compiler ghc841 --show-trace
error: while evaluating the attribute ‘propagatedBuildInputs’ of the derivation ‘text-lens-0.1.1’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘propagatedBuildInputs’ of the derivation ‘lens-4.15.4’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘buildInputs’ of the derivation ‘base-orphans-0.6’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘propagatedBuildInputs’ of the derivation ‘hspec-2.4.7’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘buildInputs’ of the derivation ‘hspec-core-2.4.7’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘buildInputs’ of the derivation ‘silently-1.2.5’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘buildInputs’ of the derivation ‘temporary-1.2.1.1’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
while evaluating the attribute ‘buildInputs’ of the derivation ‘base-compat-0.9.3’ at /home/deepfire/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
infinite recursion encountered, at undefined position
```
Am I doing something obviously wrong?

deepfire replied:
My guess is that `overrideCabal` cannot be used for those shadow attributes for some reason.. although I'm somewhat surprised, given how they are defined (same as normal, non-version-suffixed ones).

deepfire replied:
For a bit more context -- the construction I'm using is:
```
ghcOrig.override (oldArgs: {
    overrides = new: old:
    with new; ((oldArgs.overrides or (_: _: {})) new old)
    with haskellLib; with self; {
       ...
hspec-core = overrideCabal super.hspec-core_2_4_7 (drv: {
    editedCabalFile = null;
  });
         ...
    };
}
```

peti replied:
The `hspec` packages are tricky because some of their dependencies depend on `hspec` for testing --- which causes infinite loops. The easiest way to avoid that issue is to disable the check phase of those packages, like I did in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix#L46.

A better albeit more involved solution is to figure out which package is causing the recursion and to pass a dont-check-variant of that particular package to `hspec` during testing, like we do at https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/configuration-common.nix#L51-L52. We have those overrides for the default versions, but apparently we don't have them for the latest versions, which makes sense, I guess.

BTW, I'm wondering why you set 

    editedCabalFile = null;

in your override. That doesn't seem necessary to me?

deepfire replied:
The `editedCabalFile` is coming from a pre-auto-trim state of the override set, but after I have promoted the package from github-overridden (where the editedCabalFile negation is necessary) to "hackaged" -- basically intermediate state.

Investigating now how to deal with the loops -- mainly how much of the problems can be dealt with `doCheck = false`..

deepfire replied:
Done -- the use of versioned attributes is now automatic.

peti commented on pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix:
```diff
@@ -41,4 +41,1041 @@ self: super: {
+    ##       from the context: (Monad m, Monoid a)
+    src = pkgs.fetchgit {
+      url    = "https://github.com/ekmett/free";
+      rev    = "fcefc71ed302f2eaf60f020046bad392338b3109";
```
How will this override be updated when a new release version of `free` comes out that supersedes this one?

deepfire replied:
Good question! -- I've added automation to the audit script to catch the override version lagging behind upstream to the update script.  I'll update the override set tonight.

So the full answer is -- there are scripts that:
1. automatically generate set trimming proposals, and
2. audit the current override set for various invariants, like:
- Nixpkgs-supplied packages lagging behind the Hackage-based overrides 
- Upstreamed-overridden version not lagging behind upstream `master`, in terms of Cabal version number
- User-overridden version having a PR supplied and the PR _not_ being merged
- User-overridden version having the same Cabal version as upstream `master`

When I run the audit script and see any warnings, that means actions need to be taken.

peti replied:
OK, if you've automated these things, then that's a pretty good solution. :-) It would be great if you could provide a brief step-by-step write-up of how to run those scripts of yours so that I know how to run them, too.

deepfire replied:
My only concern them being such a mess, I'm somewhat ashamed to bring them to light -- this is truly a case of exploratory chaos -- in bash..

That said, I'll close my eyes and still do this -- just give me a bit more time..

deepfire replied:
I've mostly refactored the thing into a tolerable shape.

I guess a small test suite is what remains.

peti commented on pkgs/development/haskell-modules/configuration-ghc-8.4.x.nix:
```diff
@@ -41,4 +41,1041 @@ self: super: {
+    ##     • Could not deduce (Semigroup (IterT m a))
+    ##         arising from the superclasses of an instance declaration
+    ##       from the context: (Monad m, Monoid a)
+    src = pkgs.fetchgit {
```
Would it be possible to use `fetchGitHub` instead of `fetchgit`? It's a minor optimization, granted, but `fetchGitHub` downloads a single tarball rather than fetching each file individually and is therefore more efficient.

deepfire replied:
Sure, I'll do that.

deepfire replied:
Done.

peti reviewed (approved):
Great work! I'd be happy to merge this PR. Let me know when all pending stages are pushed, please.

deepfire committed:
ghc841 configuration:  overrides

deepfire commented:
@peti, I believe I have addressed your suggestions.

I can't vouch 100% on the necessity of these overrides, since they are auto-generated, and there's lots of things to keep in mind -- heuristics can only go so far.
